### PR TITLE
Improvement:  Better usability of remote screen

### PR DESCRIPTION
**Pull request by wutschel**
## Description

Closes https://github.com/xbmc/Official-Kodi-Remote-iOS/issues/573.

This PR implemented some changes in the remote functionality:

iPad Remote Screen:
- Use `UIModalPresentationFormSheet` for the remote view on iPad:
- Remote view can be brought up from anywhere via a button in the iPad toolbar.
- Remote view on iPad does not dismiss the browser screen anymore, it is shown as overlay.
- Remote view is always shown in center.
- Remote view needs to have the volume slider as hitting the volume button in toolbar will dismiss the overlay.

iPhone Remote Screen:
- Remote Position can be toggled at runtime via new toolbar button.

Persistency:
- The gesture zone mode and remote position are now saved and restored.
- The app settings for the same were removed.

Screenshots:
https://abload.de/img/simulatorscreenshot-iy8ku5.png (iPad, overlay over browser)
https://abload.de/img/simulatorscreenshot-iofkk4.png (iPad, overlay over fullscreen)
https://abload.de/img/simulatorscreenshot-ic2kvy.png (iPad, overlay with volume bar)
https://abload.de/img/bildschirmfoto2022-0195jwi.png (iPhone, top/bottom toggle)

## Summary for release notes

Improvement: Remote screen on iPad does not dismiss browser screens
Improvement: Runtime change of remote top/bottom position on iPhone
Improvement: Restore remote modes instead of app settings configuration

### CODE REVIEW

File: XBMC Remote/RemoteController.m [hunk 1]
```diff
@@ -1180,6 +1181,33 @@ - (void)hideKeyboard:(id)sender {
...
 
+#pragma mark - Persistence
+
+- (void)saveRemoteMode {
+    NSDictionary *modes = @{@"GestureView": @(isGestureViewActive),
+                            @"RemotePosition": @(positionMode)};
+    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
+    if (paths.count > 0) {
+        NSString *path = [paths[0] stringByAppendingPathComponent:@"RemoteModes.dat"];
+        [NSKeyedArchiver archiveRootObject:modes toFile:path];
```
Comment (kambala-decapitator) [inline XBMC Remote/RemoteController.m:1192]: why not using user defaults?

**Review by kambala-decapitator — CHANGES_REQUESTED** [review]
also regarding 9590d3dff37f2af097793a998a0169a969512ce4: `stting` typo, don't see a reason to mention cherry-pick

File: XBMC Remote/RemoteController.m [hunk 2]
```diff
@@ -251,19 +251,33 @@ - (void)configureView {
...
         // Calculate the maximum possible scaling for the remote
-        CGFloat scaleFactorHorizontal = STACKSCROLL_WIDTH / CGRectGetWidth(remoteControlView.frame);
-        CGFloat minViewHeight = MIN(CGRectGetWidth(UIScreen.mainScreen.fixedCoordinateSpace.bounds), CGRectGetHeight(UIScreen.mainScreen.fixedCoordinateSpace.bounds)) - REMOTE_PADDING - bottomPadding;
+        CGFloat scaleFactorHorizontal = PAD_REMOTE_WIDTH / CGRectGetWidth(remoteControlView.frame);
+        CGFloat minViewHeight = MIN(CGRectGetWidth(UIScreen.mainScreen.fixedCoordinateSpace.bounds), CGRectGetHeight(UIScreen.mainScreen.fixedCoordinateSpace.bounds)) - REMOTE_PADDING - bottomPadding - CGRectGetMaxY(volumeSliderView.frame);
         CGFloat scaleFactorVertical = minViewHeight / CGRectGetHeight(remoteControlView.frame);
         CGFloat transform = MIN(scaleFactorHorizontal, scaleFactorVertical);
 
-        CGRect frame = remoteControlView.frame;
+        frame = remoteControlView.frame;
```
[the remark this answers is not in the document]
Comment (wutschel) [reply XBMC Remote/RemoteController.m:273]: Good catch! In fact the former assignment and condition could be made much simpler.

**Comment by wutschel** [comment]
> also regarding [9590d3d](https://github.com/xbmc/Official-Kodi-Remote-iOS/commit/9590d3dff37f2af097793a998a0169a969512ce4): `stting` typo, don't see a reason to mention cherry-pick

Fixed.

File: XBMC Remote/RemoteController.m [hunk 3]
```diff
@@ -691,6 +691,10 @@ - (void)changeServerVolume {
 }
 
 - (void)touchesBegan:(NSSet*)touches withEvent:(UIEvent*)event {
+    if (IS_IPAD) {
+        // Disable gestures to move the modal remote as tbis would conflicts with the gesture zone
+        [self.presentationController.presentedView.gestureRecognizers[0] setEnabled:NO];
```
[the remark this answers is not in the document]
Comment (wutschel) [reply XBMC Remote/RemoteController.m:717]: I didn't find any other way to handle this.

File: XBMC Remote/RemoteController.m (hunk 1 above)
Comment (wutschel) [reply XBMC Remote/RemoteController.m]: You're asking the right question. With user defaults this is much less code and easier to read.

File: XBMC Remote/RemoteController.m (hunk 3 above)
[the remark this answers is not in the document]
Comment (kambala-decapitator) [reply XBMC Remote/RemoteController.m:710]: I think it's related to the fact that modern OS uses `UIModalPresentationAutomatic` style by default when showing modal VCs, try setting `modalPresentationStyle` property to `UIModalPresentationFullScreen` instead (or a different one)

File: XBMC Remote/RemoteController.m [hunk 4]
```diff
@@ -1177,27 +1177,16 @@ - (void)hideKeyboard:(id)sender {
...
 - (void)saveRemoteMode {
-    NSDictionary *modes = @{@"GestureView": @(isGestureViewActive),
-                            @"RemotePosition": @(positionMode)};
-    NSArray *paths = NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES);
-    if (paths.count > 0) {
-        NSString *path = [paths[0] stringByAppendingPathComponent:@"RemoteModes.dat"];
-        [NSKeyedArchiver archiveRootObject:modes toFile:path];
-    }
+    NSUserDefaults *userDefaults = [NSUserDefaults standardUserDefaults];
+    [userDefaults setObject:@(isGestureViewActive) forKey:@"GestureView"];
```
Comment (kambala-decapitator) [inline XBMC Remote/RemoteController.m:1181]: FYI there're dedicated `integerForKey` and `boolForKey` as well as respective setters.

Also, `GestureView` doesn't seem a good name for a boolean.

**Review by kambala-decapitator — CHANGES_REQUESTED** [review]

File: XBMC Remote/RemoteController.m (hunk 3 above)
[the remark this answers is not in the document]
Comment (wutschel) [reply XBMC Remote/RemoteController.m:710]: I tried using `UIModalPresentationFullScreen` (and others). The current way definitely looks best. We should not change this.

File: XBMC Remote/RemoteController.m (hunk 4 above)
Comment (wutschel) [reply XBMC Remote/RemoteController.m:1181]: I will apply `integerForKey` and `boolForKey` in a dedicated PR. The key name I will change to `"GestureViewEnabled"`.

File: XBMC Remote/RemoteController.m (hunk 3 above)
[the remark this answers is not in the document]
Comment (kambala-decapitator) [reply XBMC Remote/RemoteController.m:710]: ok, let's keep.

have you tested on pre-iOS 13? I feel it'd crash with array out-of-bounds. Using `gestureRecognizers.first` would be safer.

[the remark this answers is not in the document]
Comment (wutschel) [reply XBMC Remote/RemoteController.m]: > have you tested on pre-iOS 13? 

Just tested with iPad simulator using iOS 11.4 -> works. On iOS 11.4 it even works w/o this line. But on iOS 14.5 the line is required. Do we know by hard fro which iOS on this setting is needed?

> I feel it'd crash with array out-of-bounds. Using `gestureRecognizers.first` would be safer.

Changed to `gestureRecognizers.firstObject`.

[the remark this answers is not in the document]
Comment (kambala-decapitator) [reply XBMC Remote/RemoteController.m]: I believe it was introduced in iOS 13 and is directly related to UIModalPresentationAutomatic

[the remark this answers is not in the document]
Comment (kambala-decapitator) [reply XBMC Remote/RemoteController.m]: ok, let's keep without condition

**Review by kambala-decapitator — APPROVED** [review]

**Comment by wutschel** [comment]
Squashed and rebased to master.

**Pull request merged**